File: Freedom/微信/我/Setting/WXCommonSettingViewController.m
//  WXCommonSettingViewController.m
//  Freedom
// Created by Super
#import "WXCommonSettingViewController.h"
#import "WXMessageManager.h"
#import "WXChatViewController.h"
#import "WXBgSettingViewController.h"
#import "WXMyExpressionViewController.h"
#import "WXChatTableViewController.h"
#import "WXUserHelper.h"
#import "WXBaseViewController.h"
#import "NSFileManager+expanded.h"
#import "WXModes.h"
@interface WXChatFontSettingView : UIView
@property (nonatomic, assign) CGFloat curFontSize;
@property (nonatomic, copy) void (^fontSizeChangeTo)(CGFloat size);
@end
@interface WXChatFontSettingView ()
@property (nonatomic, strong) UILabel *miniFontLabel;
@property (nonatomic, strong) UILabel *maxFontLabel;
@property (nonatomic, strong) UILabel *standardFontLabel;
@property (nonatomic, strong) UISlider *slider;
@end
@implementation WXChatFontSettingView
- (id)initWithFrame:(CGRect)frame{
    if (self = [super initWithFrame:frame]) {
        [self setBackgroundColor:[UIColor whiteColor]];
        [self addSubview:self.miniFontLabel];
        [self addSubview:self.maxFontLabel];
        [self addSubview:self.standardFontLabel];
        [self addSubview:self.slider];
        [self p_addMasonry];
    }
    return self;
}
- (void)setCurFontSize:(CGFloat)curFontSize{
    _curFontSize = curFontSize;
    [self.slider setValue:curFontSize];
}
#pragma mark - Event Response -
- (void)sliderValueChanged:(UISlider *)sender{
    NSInteger value = (NSInteger)sender.value;
    value = ((sender.value - value) > 0.5 ? value + 1 : value);
    if (value == (NSInteger)_curFontSize) {
        return;
    }
    _curFontSize = value;
    if (self.fontSizeChangeTo) {
        self.fontSizeChangeTo(value);
    }
}
#pragma mark - Private Methods -
- (void)p_addMasonry{
    [self.miniFontLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.mas_equalTo(self.slider.mas_left);
        make.bottom.mas_equalTo(self.slider.mas_top).mas_offset(-6);
    }];
    
    [self.maxFontLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.right.mas_equalTo(self.slider.mas_right);
        make.bottom.mas_equalTo(self.miniFontLabel);
    }];
    
    [self.standardFontLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.bottom.mas_equalTo(self.miniFontLabel);
        make.left.mas_equalTo(self.miniFontLabel.mas_right).mas_equalTo(40);
    }];
    
    [self.slider mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerX.mas_equalTo(self);
        make.bottom.mas_equalTo(self.mas_bottom).mas_offset(-35);
        make.width.mas_equalTo(self).multipliedBy(0.8);
        make.height.mas_equalTo(40);
    }];
}
#pragma mark - Getter -
- (UILabel *)miniFontLabel{
    if (_miniFontLabel == nil) {
        _miniFontLabel = [[UILabel alloc] init];
        [_miniFontLabel setFont:[UIFont systemFontOfSize:15]];
        [_miniFontLabel setText:@"A"];
    }
    return _miniFontLabel;
}
- (UILabel *)maxFontLabel{
    if (_maxFontLabel == nil) {
        _maxFontLabel = [[UILabel alloc] init];
        [_maxFontLabel setFont:[UIFont systemFontOfSize:20]];
        [_maxFontLabel setText:@"A"];
    }
    return _maxFontLabel;
}
- (UILabel *)standardFontLabel{
    if (_standardFontLabel == nil) {
        _standardFontLabel = [[UILabel alloc] init];
        [_standardFontLabel setFont:[UIFont systemFontOfSize:16]];
        [_standardFontLabel setText:@"标准"];
    }
    return _standardFontLabel;
}
- (UISlider *)slider{
    if (_slider == nil) {
        _slider = [[UISlider alloc] init];
        [_slider setMinimumValue:15];
        [_slider setMaximumValue:20];
        [_slider addTarget:self action:@selector(sliderValueChanged:) forControlEvents:UIControlEventValueChanged];
    }
    return _slider;
}
@end
@interface WXChatFontViewController : WXBaseViewController
@end
@interface WXChatFontViewController ()
@property (nonatomic, strong) WXChatTableViewController *chatTVC;
@property (nonatomic, strong) WXChatFontSettingView *chatFontSettingView;
@end
@implementation WXChatFontViewController
- (void)viewDidLoad{
    [super viewDidLoad];
    [self.navigationItem setTitle:@"字体大小"];
    
    [self.view addSubview:self.chatTVC.view];
    [self addChildViewController:self.chatTVC];
    [self.view addSubview:self.chatFontSettingView];
    [self p_addMasonry];
    
    __weak typeof(self) weakSelf = self;
    [self.chatFontSettingView setFontSizeChangeTo:^(CGFloat size) {
        [[NSUserDefaults standardUserDefaults] setDouble:size forKey:@"CHAT_FONT_SIZE"];
        weakSelf.chatTVC.data = [weakSelf p_chatTVCData];
        [weakSelf.chatTVC.tableView reloadData];
    }];
    CGFloat size = [[NSUserDefaults standardUserDefaults] doubleForKey:@"CHAT_FONT_SIZE"];
    [self.chatFontSettingView setCurFontSize:size];
    self.chatTVC.data = [self p_chatTVCData];
    [self.chatTVC.tableView reloadData];
}
#pragma mark - Private Methods -
- (void)p_addMasonry{
    [self.chatTVC.view mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.and.left.and.right.mas_equalTo(self.view);
        make.bottom.mas_equalTo(self.chatFontSettingView.mas_top);
    }];
    
    [self.chatFontSettingView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.and.right.and.bottom.mas_equalTo(self.view);
        make.height.mas_equalTo(self.chatFontSettingView.mas_width).multipliedBy(0.4);
    }];
}
- (NSMutableArray *)p_chatTVCData{
    WXTextMessage *message = [[WXTextMessage alloc] init];
    message.fromUser = [WXUserHelper sharedHelper].user;
    message.messageType = TLMessageTypeText;
    message.ownerTyper = TLMessageOwnerTypeSelf;
    message.content[@"text"] = @"预览字体大小";
    
    WXUser *user = [[WXUser alloc] init];
    user.avatarPath = @"AppIcon";
    NSString *path = [NSFileManager pathUserAvatar:@"AppIcon"];
    if (![[NSFileManager defaultManager] isExecutableFileAtPath:path]) {
        NSString *iconPath = [[[[NSBundle mainBundle] infoDictionary] valueForKeyPath:@"CFBundleIcons.CFBundlePrimaryIcon.CFBundleIconFiles"] lastObject];
        UIImage *image = [UIImage imageNamed:iconPath];
        NSData *data = (UIImagePNGRepresentation(image) ? UIImagePNGRepresentation(image) :UIImageJPEGRepresentation(image, 1));
        [[NSFileManager defaultManager] createFileAtPath:path contents:data attributes:nil];
    }
    
    
    WXTextMessage *message1 = [[WXTextMessage alloc] init];
    message1.fromUser = user;
    message1.messageType = TLMessageTypeText;
    message1.ownerTyper = TLMessageOwnerTypeFriend;
    message1.content[@"text"] = @"拖动下面的滑块，可设置字体大小";
    WXTextMessage *message2 = [[WXTextMessage alloc] init];
    message2.fromUser = user;
    message2.messageType = TLMessageTypeText;
    message2.ownerTyper = TLMessageOwnerTypeFriend;
    message2.content[@"text"] = @"设置后，会改变聊天页面的字体大小。后续将支持更改菜单、朋友圈的字体修改。";
    
    NSMutableArray *data = [[NSMutableArray alloc] initWithObjects:message, message1, message2, nil];
    return data;
}
#pragma mark - Getter -
- (WXChatTableViewController *)chatTVC{
    if (_chatTVC == nil) {
        _chatTVC = [[WXChatTableViewController alloc] init];
        [_chatTVC setDisablePullToRefresh:YES];
        [_chatTVC setDisableLongPressMenu:YES];
    }
    return _chatTVC;
}
- (WXChatFontSettingView *)chatFontSettingView{
    if (_chatFontSettingView == nil) {
        _chatFontSettingView = [[WXChatFontSettingView alloc] init];
    }
    return _chatFontSettingView;
}
@end
@implementation WXCommonSettingHelper
+ (NSMutableArray *)chatBackgroundSettingData{
    WXSettingItem *select = [WXSettingItem createItemWithTitle:(@"选择背景图")];
    WXSettingGroup *group1 = TLCreateSettingGroup(nil, nil, @[select]);
    
    WXSettingItem *album = [WXSettingItem createItemWithTitle:(@"从手机相册中选择")];
    WXSettingItem *camera = [WXSettingItem createItemWithTitle:(@"拍一张")];
    WXSettingGroup *group2 = TLCreateSettingGroup(nil, nil, (@[album, camera]));
    
    WXSettingItem *toAll = [WXSettingItem createItemWithTitle:(@"将背景应用到所有聊天场景")];
    toAll.type = TLSettingItemTypeTitleButton;
    WXSettingGroup *group3 = TLCreateSettingGroup(nil, nil, @[toAll]);
    
    NSMutableArray *data = [[NSMutableArray alloc] initWithObjects:group1, group2, group3, nil];
    return data;
}
- (id) init{
    if (self = [super init]) {
        self.commonSettingData = [[NSMutableArray alloc] init];
        [self p_initTestData];
    }
    return self;
}
- (void) p_initTestData{
    WXSettingItem *item1 = [WXSettingItem createItemWithTitle:(@"多语言")];
    WXSettingGroup *group1 = TLCreateSettingGroup(nil, nil, @[item1]);
    
    WXSettingItem *item2 = [WXSettingItem createItemWithTitle:(@"字体大小")];
    WXSettingItem *item3 = [WXSettingItem createItemWithTitle:(@"聊天背景")];
    WXSettingItem *item4 = [WXSettingItem createItemWithTitle:(@"我的表情")];
    WXSettingItem *item5 = [WXSettingItem createItemWithTitle:(@"朋友圈小视频")];
    WXSettingGroup *group2 = TLCreateSettingGroup(nil, nil, (@[item2, item3, item4, item5]));
    
    WXSettingItem *item6 = [WXSettingItem createItemWithTitle:(@"听筒模式")];
    item6.type = TLSettingItemTypeSwitch;
    WXSettingGroup *group3 = TLCreateSettingGroup(nil, nil, @[item6]);
    
    WXSettingItem *item7 = [WXSettingItem createItemWithTitle:(@"功能")];
    WXSettingGroup *group4 = TLCreateSettingGroup(nil, nil, @[item7]);
    
    WXSettingItem *item8 = [WXSettingItem createItemWithTitle:(@"聊天记录迁移")];
    WXSettingItem *item9 = [WXSettingItem createItemWithTitle:(@"清理微信存储空间")];
    WXSettingGroup *group5 = TLCreateSettingGroup(nil, nil, (@[item8, item9]));
    
    WXSettingItem *item10 = [WXSettingItem createItemWithTitle:(@"清空聊天记录")];
    item10.type = TLSettingItemTypeTitleButton;
    WXSettingGroup *group6 = TLCreateSettingGroup(nil, nil, @[item10]);
    
    [self.commonSettingData addObjectsFromArray:@[group1, group2, group3, group4, group5, group6]];
}
@end
@interface WXCommonSettingViewController ()
@property (nonatomic, strong) WXCommonSettingHelper *helper;
@end
@implementation WXCommonSettingViewController
- (void)viewDidLoad {
    [super viewDidLoad];
    [self.navigationItem setTitle:@"通用"];
    
    self.helper = [[WXCommonSettingHelper alloc] init];
    self.data = self.helper.commonSettingData;
}
#pragma mark - Delegate -
//MARK: UITableViewDelegate
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    WXSettingItem *item = [self.data[indexPath.section] objectAtIndex:indexPath.row];
    if ([item.title isEqualToString:@"字体大小"]) {
        WXChatFontViewController *chatFontVC = [[WXChatFontViewController alloc] init];
        [self setHidesBottomBarWhenPushed:YES];
        [self.navigationController pushViewController:chatFontVC animated:YES];
    }else if ([item.title isEqualToString:@"聊天背景"]) {
        WXBgSettingViewController *chatBGSettingVC = [[WXBgSettingViewController alloc] init];
        [self setHidesBottomBarWhenPushed:YES];
        [self.navigationController pushViewController:chatBGSettingVC animated:YES];
    }else if ([item.title isEqualToString:@"我的表情"]) {
        WXMyExpressionViewController *myExpressionVC = [[WXMyExpressionViewController alloc] init];
        [self setHidesBottomBarWhenPushed:YES];
        [self.navigationController pushViewController:myExpressionVC animated:YES];
    }else if ([item.title isEqualToString:@"清空聊天记录"]) {
        [self showAlerWithtitle:@"将删除所有个人和群的聊天记录。" message:nil style:UIAlertControllerStyleActionSheet ac1:^UIAlertAction *{
            return [UIAlertAction actionWithTitle:@"清空聊天记录" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                    [[WXMessageManager sharedInstance] deleteAllMessages];
                    [[WXChatViewController sharedChatVC] resetChatVC];
            }];
        } ac2:^UIAlertAction *{
            return [UIAlertAction actionWithTitle:@"取消" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {

            }];
        } ac3:nil completion:nil];
    }
    
    [tableView deselectRowAtIndexPath:indexPath animated:NO];
}
@end

File: Freedom/微信/我/Setting/WXCommonSettingViewController.swift
//
//  WXCommonSettingViewController.swift
//  Freedom
import SnapKit
//import XExtension
import Foundation
class WXChatFontSettingView: UIView {
    var curFontSize: Float = 0.0 {
        didSet {
            slider.value = curFontSize
        }
    }
    var fontSizeChangeTo: ((_ size: Float) -> Void)?
    lazy var miniFontLabel: UILabel = {
        let miniFontLabel = UILabel()
        miniFontLabel.font = UIFont.normal
        miniFontLabel.text = "A"
        return miniFontLabel
    }()
    lazy var maxFontLabel: UILabel =  {
        let maxFontLabel = UILabel()
        maxFontLabel.font = UIFont.systemFont(ofSize: 20)
        maxFontLabel.text = "A"
        return maxFontLabel
    }()
    lazy var standardFontLabel: UILabel = {
        let standardFontLabel = UILabel()
        standardFontLabel.font = UIFont.big
        standardFontLabel.text = "标准"
        return standardFontLabel
    }()
    lazy var slider: UISlider = {
        let slider = UISlider()
        slider.minimumValue = 15
        slider.maximumValue = 20
        slider.addTarget(self, action: #selector(self.sliderValueChanged(_:)), for: .valueChanged)
        return slider
    }()
    override init(frame: CGRect) {
        super.init(frame: frame)
        backgroundColor = UIColor.whitex
        addSubview(miniFontLabel)
        addSubview(maxFontLabel)
        addSubview(standardFontLabel)
        addSubview(slider)
        miniFontLabel.snp.makeConstraints { (make) in
            make.left.equalTo(self.slider.snp.left)
            make.bottom.equalTo(self.slider.snp.top).offset(-6)
        }
        maxFontLabel.snp.makeConstraints { (make) in
            make.right.equalTo(self.slider.snp.right)
            make.bottom.equalTo(self.miniFontLabel)
        }
        standardFontLabel.snp.makeConstraints { (make) in
            make.bottom.equalTo(self.miniFontLabel)
            make.left.equalTo(self.miniFontLabel.snp.right).offset(40)
        }
        slider.snp.makeConstraints { (make) in
            make.centerX.equalToSuperview()
            make.bottom.equalTo(self.snp.bottom).offset(-35)
            make.width.equalToSuperview().multipliedBy(0.8)
            make.height.equalTo(40)
        }
    }
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    @objc func sliderValueChanged(_ sender: UISlider) {
        var value = sender.value
        value = (sender.value - value) > 0.5 ? value + 1 : value
        if value == curFontSize { return }
        curFontSize = value
        fontSizeChangeTo?(value)
    }
}
class WXChatFontViewController: WXBaseViewController {
    lazy var chatTVC: WXChatTableViewController = {
        let chatTVC = WXChatTableViewController()
        chatTVC.disablePullToRefresh = true
        chatTVC.disableLongPressMenu = true
        return chatTVC
    }()
    var chatFontSettingView = WXChatFontSettingView()
    override func viewDidLoad() {
        super.viewDidLoad()
        navigationItem.title = "字体大小"
        view.addSubview(chatTVC.view)
        addChild(chatTVC)
        view.addSubview(chatFontSettingView)
        chatFontSettingView.fontSizeChangeTo = { size in
            UserDefaults.standard.set(Double(size), forKey: "CHAT_FONT_SIZE")
            self.chatTVC.data = self.p_chatTVCData()
            self.chatTVC.tableView.reloadData()
        }
        let size = Float(UserDefaults.standard.double(forKey: "CHAT_FONT_SIZE"))
        chatFontSettingView.curFontSize = size
        chatTVC.data = p_chatTVCData()
        chatTVC.tableView.reloadData()
        chatTVC.view.snp.makeConstraints { (make) in
            make.top.left.right.equalTo(self.view)
            make.bottom.equalTo(self.chatFontSettingView.snp.top)
        }
        chatFontSettingView.snp.makeConstraints { (make) in
            make.left.right.bottom.equalTo(self.view)
            make.height.equalTo(self.chatFontSettingView.snp.width).multipliedBy(0.4)
        }
    }
    func p_chatTVCData() -> [WXTextMessage] {
        let message = WXTextMessage()
        message.friendID = WXUserHelper.shared.user.userID
        message.messageType = .text
        message.ownerTyper = .own
        message.text = "预览字体大小"
        let user = WXUser()
        user.avatarPath = "AppIcon"
        let path = FileManager.pathUserAvatar("AppIcon")
        if !FileManager.default.isExecutableFile(atPath: path) {
            let iconPath = Bundle.main.infoDictionary?["CFBundleIcons.CFBundlePrimaryIcon.CFBundleIconFiles"] as! String
            let image = iconPath.image
            var data = image.pngData() != nil ? image.pngData() : image.jpegData(compressionQuality: 1)
            FileManager.default.createFile(atPath: path, contents: data, attributes: nil)
        }
        let message1 = WXTextMessage()
        message1.friendID = user.userID ?? ""
        message1.messageType = .text
        message1.ownerTyper = .friend
        message1.text = "拖动下面的滑块，可设置字体大小"
        let message2 = WXTextMessage()
        message2.friendID = user.userID ?? ""
        message2.messageType = .text
        message2.ownerTyper = .friend
        message2.text = "设置后，会改变聊天页面的字体大小。后续将支持更改菜单、朋友圈的字体修改。"
        var data = [message, message1, message2]
        return data
    }

}
class WXCommonSettingHelper: NSObject {
    var commonSettingData: [WXSettingGroup] = []
    class func chatBackgroundSettingData() -> [WXSettingGroup] {
        let select = WXSettingItem("选择背景图")
        let group1: WXSettingGroup = WXSettingGroup(nil, nil, [select])
        let album = WXSettingItem("从手机相册中选择")
        let camera = WXSettingItem("拍一张")
        let group2: WXSettingGroup = WXSettingGroup(nil, nil, ([album, camera]))
        let toAll = WXSettingItem("将背景应用到所有聊天场景")
        toAll.type = .titleButton
        let group3: WXSettingGroup = WXSettingGroup(nil, nil, [toAll])
        return [group1, group2, group3]
    }
    func p_initTestData() {
        let item1 = WXSettingItem("多语言")
        let group1: WXSettingGroup = WXSettingGroup(nil, nil, [item1])
        let item2 = WXSettingItem("字体大小")
        let item3 = WXSettingItem("聊天背景")
        let item4 = WXSettingItem("我的表情")
        let item5 = WXSettingItem("朋友圈小视频")
        let group2: WXSettingGroup = WXSettingGroup(nil, nil, ([item2, item3, item4, item5]))
        let item6 = WXSettingItem("听筒模式")
        item6.type = .switchBtn
        let group3: WXSettingGroup = WXSettingGroup(nil, nil, [item6])
        let item7 = WXSettingItem("功能")
        let group4: WXSettingGroup = WXSettingGroup(nil, nil, [item7])
        let item8 = WXSettingItem("聊天记录迁移")
        let item9 = WXSettingItem("清理微信存储空间")
        let group5: WXSettingGroup = WXSettingGroup(nil, nil, ([item8, item9]))
        let item10 = WXSettingItem("清空聊天记录")
        item10.type = .titleButton
        let group6: WXSettingGroup = WXSettingGroup(nil, nil, [item10])
        commonSettingData.append(contentsOf: [group1, group2, group3, group4, group5, group6])
    }
    override init() {
        super.init()
        p_initTestData()
    }
}
class WXCommonSettingViewController: WXSettingViewController {
    var helper = WXCommonSettingHelper()
    override func viewDidLoad() {
        super.viewDidLoad()
        navigationItem.title = "通用"
        data = helper.commonSettingData
    }
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let item = data[indexPath.section].items[indexPath.row]
        if (item.title == "字体大小") {
            let chatFontVC = WXChatFontViewController()
            navigationController?.pushViewController(chatFontVC, animated: true)
        } else if (item.title == "聊天背景") {
            let chatBGSettingVC = WXBgSettingViewController()
            navigationController?.pushViewController(chatBGSettingVC, animated: true)
        } else if (item.title == "我的表情") {
            let myExpressionVC = WXMyExpressionViewController()
            navigationController?.pushViewController(myExpressionVC, animated: true)
        } else if (item.title == "清空聊天记录") {
            let alert = UIAlertController("将删除所有个人和群的聊天记录。", "", T1: "清空聊天记录", T2: "取消", confirm1: {
                _ = WXMessageHelper.shared.deleteAllMessages()
                WXChatViewController.shared.resetChatVC()
            }, confirm2: {

            })
            self.present(alert, animated: true, completion: nil)
        }
        tableView.deselectRow(at: indexPath, animated: false)
    }
}
